File: 2023/17.rb
require 'algorithms'
input = File.read(File.basename(__FILE__).gsub('rb', 'input')).strip.split("\n").map { _1.chars.map(&:to_i) }

def dijkstra(input, start, finish, min_steps, max_steps)
  dist = Hash.new { Float::INFINITY }
  visited = {}
  unvisited = Containers::MinHeap.new
  unvisited << [0, start, []]
  dist[(start + []).to_s] = 0

  until unvisited.empty?
    cost, current, dir = unvisited.pop

    next if visited[(current + dir).to_s]

    visited[(current + dir).to_s] = true

    [[0, 1], [1, 0], [0, -1], [-1, 0]].each do |new_dir|
      new_cost = 0

      if new_dir == dir || (new_dir[0] == dir[0] && new_dir[1] != dir[1]) || (new_dir[0] != dir[0] && new_dir[1] == dir[1])
        next
      end

      (1..max_steps).each do |stretch|
        step = [current[0] + new_dir[0] * stretch, current[1] + new_dir[1] * stretch]
        next unless step[0] >= 0 && step[0] < input.size && step[1] >= 0 && step[1] < input[0].size

        new_cost += input[step[0]][step[1]]
        next if stretch < min_steps

        if new_cost + cost < dist[(step + new_dir).to_s]
          dist[(step + new_dir).to_s] = new_cost + cost
          unvisited << [dist[(step + new_dir).to_s], step, new_dir]
        end
      end
    end
  end
  [[0, 1], [1, 0], [0, -1], [-1, 0]].map { dist[(finish + _1).to_s] }.min
end

p dijkstra(input, [0, 0], [input.size - 1, input[0].size - 1], 1, 3)
p dijkstra(input, [0, 0], [input.size - 1, input[0].size - 1], 4, 10)

File: 2023/17.py
from heapq import heappop, heappush
from collections import defaultdict

input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n")
input = [[int(x) for x in y] for y in input]


def dijkstra(start, finish, min_steps, max_steps):
    dist = defaultdict(lambda: 999_999)
    visited = {}
    unvisited = [[0, start, []]]
    dist[repr(start + [])] = 0

    while unvisited:
        cost, current, dir = heappop(unvisited)

        if repr(current + dir) in visited:
            continue

        visited[repr(current + dir)] = True

        for new_dir in [[0, 1], [1, 0], [0, -1], [-1, 0]]:
            new_cost = 0

            if (
                new_dir == dir
                or (len(dir) and new_dir[0] == dir[0] and new_dir[1] != dir[1])
                or (len(dir) and new_dir[0] != dir[0] and new_dir[1] == dir[1])
            ):
                continue

            for stretch in range(1, max_steps + 1):
                step = [
                    current[0] + new_dir[0] * stretch,
                    current[1] + new_dir[1] * stretch,
                ]
                if (
                    step[0] < 0
                    or step[0] >= len(input)
                    or step[1] < 0
                    or step[1] >= len(input[0])
                ):
                    continue

                new_cost += input[step[0]][step[1]]
                if stretch < min_steps:
                    continue

                if new_cost + cost < dist[repr(step + new_dir)]:
                    dist[repr(step + new_dir)] = new_cost + cost
                    heappush(unvisited, [new_cost + cost, step, new_dir])

    return min([dist[repr(finish + x)] for x in [[0, 1], [1, 0], [0, -1], [-1, 0]]])


print(dijkstra([0, 0], [len(input) - 1, len(input[0]) - 1], 1, 3))
print(dijkstra([0, 0], [len(input) - 1, len(input[0]) - 1], 4, 10))
